### PR TITLE
Moving average (Simple and Weighted) Implementation

### LINKED ISSUE #14: Implement moving average
In the analysis sub-module add weighted and simple moving average implementation. Make use of Pandas.

### PR DESCRIPTION
**Pull request by kushkansara02**
#### The implementation of Issue #14 
Implemented two functions in moving_average.py within the analysis submodule. These functions take in a Pandas dataframe and some other specifications as parameters, and return another dataframe. The dataframe will have the moving average of the closing prices (configurable to be open, high, low as well).
#### Testing
I have also added some tests for the moving average functions using pytest to ensure they are working correctly. Please let me know if there are any errors or things I can change.

### CODE REVIEW

File: marketools/analysis/moving_average.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                          min_periods: int = 0):
+    """
+    Returns Pandas Dataframe with the Simple Moving Average for Open, High, Low, Close. NaN will be placed for days
+    where there is not enough previous data (depends on min_periods)
+
+    Parameters
+    ----------
+    ohlc : pandas.DataFrame
+        DataFrame with OHLC data
+    config: list
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: missing space after config

File: marketools/analysis/moving_average.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import pandas as pd
+import numpy as np
+
+
+def simple_moving_average(ohlc: pd.DataFrame,
+                          config=["Close"],
+                          window: int = 10,
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: I think 15 is more common, use 15 please

File: marketools/analysis/moving_average.py (hunk 1 above)
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: 'config' doesn't say much. I would suggest to name it 'price' and make a string from it. then the function calculates SMA only for one price (Close by default) and returns pandas.Series (preferably)

File: marketools/analysis/moving_average.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    # implement moving average for the OHLC columns
+    for col in config:
+        output[col] = ohlc[col].rolling(window=window, min_periods=min_periods).mean()
+
+    return output
+
+
+def weighted_moving_average(ohlc: pd.DataFrame,
+                            config=["Close"],
+                            window: int = 10):
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: I think here 10 is fine, but maybe we should be consistent and use 15 too. I let you decide.

File: marketools/analysis/moving_average.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    Returns Pandas Dataframe with the Weighted Moving Average for Open, High, Low, Close. NaN will be placed for days
+    where there is not enough previous data
+
+    Parameters
+    ----------
+    ohlc : pandas.DataFrame
+        DataFrame with OHLC data
+    config: list
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: same comments as for SMA

File: marketools/analysis/moving_average.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        List that specifies on which aspects of OHLC moving average will be applied.
+    window : int
+        Size of the moving window. This is the number of observations used for calculating the statistic
+
+    Returns
+    -------
+    pandas.DataFrame
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: should return pandas.Series

File: marketools/analysis/moving_average.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
+import pandas as pd
+import numpy as np
+
+
+def simple_moving_average(ohlc: pd.DataFrame,
+                          config=["Close"],
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: add type hint

File: marketools/analysis/moving_average.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    # implement moving average for the OHLC columns
+    for col in config:
+        output[col] = ohlc[col].rolling(window=window, min_periods=min_periods).mean()
+
+    return output
+
+
+def weighted_moving_average(ohlc: pd.DataFrame,
+                            config=["Close"],
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: add type hint

**Review by AlbertRtk — CHANGES_REQUESTED** [review]
Some minor corrections, logic is ok.
Optional: could you add exponential_moving_average?

**Comment by kushkansara02** [comment]
I changed up the implementation according to what you said, and also added EMA. Please check it over again

File: marketools/analysis/moving_average.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    price : str
+        The price on which EMA will be calculated (Close by default)
+
+    Returns
+    -------
+    pandas.Series
+    """
+
+    # implement exponential moving average
+    output = ohlc[price].ewm(alpha=alpha, adjust=False).mean()
```
Comment (AlbertRtk) [inline marketools/analysis/moving_average.py]: Please, use span instead of alpha here. Then it's ok to merge
(remember to update docstring)

**Review by AlbertRtk — COMMENTED** [review]
One last chance and it is ok.
FYI: I changed the license from MIT to BSD-3-Clause License. Hope it's ok?

**Comment by AlbertRtk** [comment]
One last change and we can merge it.
FYI I changed license from MIT to BSD-3-Clause License. Hope you ok with it?
Regards,
Albert Ratajczak


śr., 2 gru 2020 o 20:15 kushkansara02 <notifications@github.com> napisał(a):

**Comment by kushkansara02** [comment]
Done! changed alpha to span in the code and docstring

**Review by AlbertRtk — APPROVED** [review]

**Comment by AlbertRtk** [comment]
Merged, thank you :-)
Regards,
Albert Ratajczak


śr., 2 gru 2020 o 21:27 kushkansara02 <notifications@github.com> napisał(a):

**Pull request merged**